### PR TITLE
Remove redundant code from T5 encoder mask creation

### LINKED ISSUE #27211: T5 automatic creation of the `encoder_attention_mask`
Hi there,

I am going through the T5 implementation to try to replicate it and noticed that the two code blocks reported below do the same thing when creating the `encoder_attention_mask`. Is this wanted?

https://github.com/huggingface/transformers/blob/037fb7d0e1086146612a716ef914305160134e9c/src/transformers/models/t5/modeling_t5.py#L1029-L1033

https://github.com/huggingface/transformers/blob/037fb7d0e1086146612a716ef914305160134e9c/src/transformers/models/t5/modeling_t5.py#L1045-L1049

### PR DESCRIPTION
**Pull request by pietrolesci**
# What does this PR do?

Removes redundant code in the creation of the encoder attention mask for T5 as discussed in https://github.com/huggingface/transformers/issues/27211.

@amyeroberts

### CODE REVIEW

**Review by amyeroberts — COMMENTED** [review]
Thanks for fixing this! 

For the code quality checks could you: 
* Run `pip install -e .[quality]` to install the required black and ruff versions
* Run `make fix-copies` and `make fixup` and then push the applied changes to the branch

**Comment by pietrolesci** [comment]
Done :)

File: src/transformers/models/t5/modeling_t5.py [hunk 1]
```diff
@@ -1024,18 +1024,13 @@ def forward(
             if not self.is_decoder:
                 raise ValueError(f"`use_cache` can only be set to `True` if {self} is used as a decoder")
 
-        if attention_mask is None:
-            attention_mask = torch.ones(batch_size, mask_seq_length, device=inputs_embeds.device)
-        if self.is_decoder and encoder_attention_mask is None and encoder_hidden_states is not None:
-            encoder_seq_length = encoder_hidden_states.shape[1]
-            encoder_attention_mask = torch.ones(
-                batch_size, encoder_seq_length, device=inputs_embeds.device, dtype=torch.long
```
Comment (amyeroberts) [inline src/transformers/models/t5/modeling_t5.py:1032]: Just one last thing - we need to add `dtype=torch.long` on L1044 to make sure the mask is created with the same dtype as before.

**Review by amyeroberts — COMMENTED** [review]
Thanks again! One final comment. After that I think we're good to merge :)

File: src/transformers/models/t5/modeling_t5.py (hunk 1 above)
Comment (pietrolesci) [reply src/transformers/models/t5/modeling_t5.py:1032]: Hi @amyeroberts, thanks for catching this and for your comments. Worked on this very quickly yesterday (i.e., sloppy PR 🙃). Went over it again now and added typecasting where appropriate

File: src/transformers/models/mt5/modeling_mt5.py [hunk 2]
```diff
@@ -997,18 +997,13 @@ def forward(
...
-            encoder_attention_mask = torch.ones(
-                batch_size, encoder_seq_length, device=inputs_embeds.device, dtype=torch.long
-            )
-
         # initialize past_key_values with `None` if past does not exist
         if past_key_values is None:
             past_key_values = [None] * len(self.block)
 
+        if attention_mask is None:
+            attention_mask = torch.ones(batch_size, mask_seq_length, device=inputs_embeds.device, dtype=torch.long)
```
Comment (amyeroberts) [inline src/transformers/models/mt5/modeling_mt5.py:1005]: Sorry - I wasn't clear enough on may last comment. For the creation of `attention_mask` and `encoder_attention_mask` we need both to have the exact same instantiation as before to avoid any unintended changes to current behaviour. So for `attention_mask` we don't want to have `dtype=torch.long`. 

```suggestion
            attention_mask = torch.ones(batch_size, mask_seq_length, device=inputs_embeds.device)
```

**Review by amyeroberts — COMMENTED** [review]
Almost there - just one small change. Thanks for your patience and iterating on this!

**Review by amyeroberts — APPROVED** [review]
LGTM!

File: src/transformers/models/mt5/modeling_mt5.py (hunk 2 above)
Comment (amyeroberts) [reply src/transformers/models/mt5/modeling_mt5.py]: The returned dtype from the tokenizer _should_ be `torch.int64` i.e. long. If it's float, then I'd be interested to know the checkpoint as that shouldn't be the case! 

This means of course that this logic is a bit inconsistent. However, most of the time the models are called with inputs prepared by the tokenizer, so this code path is never touched. 

Unfortunately, even though setting `dtype=torch.long` is more correct, part of managing the codebase is handling backwards compatibility. I actually don't think this change would cause issues, however it's very hard to know especially for popular models like T5 and tracking numerical issues is a pain 😅

Comment (pietrolesci) [reply src/transformers/models/mt5/modeling_mt5.py]: Fully understand and agree that backward compatibility is essential!

Even though feeding `input_ids`, `labels`, and `attention_mask` to the forward method of T5 is usually enough, sometimes it's essential to have more control. For example, to compute the likelihood of continuations, i.e. `-logp(continuation | context)` for evaluation purposes (e.g., the OpenLLM leaderboard). That's why I started looking into this code :)

The backstory is that I was trying to understand what the authors of the T-Few paper were doing. To compute `logp(continuation | context)` for T5, it is essential to be able to cache the encoder outputs (i.e., the encoded context) and pass the continuations through the model to collect the loglikelihoods. In this case, you need to pass `decoder_input_ids` yourself. tl;dr: it felt strange to me how the authors computed the `decoder_attention_mask` (see https://github.com/r-three/t-few/issues/32) and to try to understand their code I started to investigate what was done by default in T5 :)
